torch/distributed/_tensor/sharding_prop.py

```diff
@@ -45,12 +38,77 @@ def register_sharding_prop_rule(
     def register_op_strategy(
         self,
         op_overload: OpOverload,
-        rule_func: Callable[[Node, DeviceMesh, Dict[Node, StrategyType]], StrategyType],
+        strategy_func: Callable[[DeviceMesh, OpSchema], StrategyType],
     ):
         """
         Register a sharding strategy generator for an operator.
         """
-        self.op_strategy_funcs[op_overload] = rule_func
+        self.op_strategy_funcs[op_overload] = strategy_func
+
+    def _propagate_tensor_meta(self, op_schema: OpSchema) -> object:
+        """
+        Propagate the tensor metadata, it could either return a TensorMeta
+        or a list/tuple of TensorMetas
+        """
+        # special case op list, we don't need to propagate for local
+        # scalar. TODO: figure out a better way to handle this
+        skip_prop_list = [
+            torch.ops.aten._local_scalar_dense.default,
+            torch.ops.aten.equal.default,
+            torch.ops.aten.is_same_size.default,
+        ]
+        if op_schema.op in skip_prop_list:
+            return None
+
+        # NOTE: We must call the tracing in fake tensor mode so that it
+        # avoids materializing memory
+        with FakeTensorMode():
+            fake_args = op_schema.gen_fake_args()
+            fake_kwargs = op_schema.gen_fake_kwargs()
+            fake_out = op_schema.op(*fake_args, **fake_kwargs)
+
+        if isinstance(fake_out, torch.Tensor):
+            return TensorMeta(
+                shape=fake_out.shape, stride=fake_out.stride(), dtype=fake_out.dtype
+            )
+
+        elif isinstance(fake_out, (tuple, list)):
+            tensor_meta_list = []
+            for i, fake_out_item in enumerate(fake_out):
+                if isinstance(fake_out_item, torch.Tensor):
+                    tensor_meta_list.append(
+                        TensorMeta(
+                            shape=fake_out_item.shape,
+                            stride=fake_out_item.stride(),
+                            dtype=fake_out_item.dtype,
+                        )
+                    )
+            return (
+                tuple(tensor_meta_list)
+                if isinstance(fake_out, tuple)
+                else tensor_meta_list
+            )
+        else:
+            # if fake is not a tensor or tuple of tensor, return as none
+            return None
+
+    def _wrap_output_spec_tensor_meta(
+        self, output_spec: OutputSpecType, output_tensor_meta: object
```

Comment:
Does this mean `output_spec` is optional?

Reply:
I think it probably won't be None. I'll submit a follow up PR to do assertion here instead.